### DISCUSSION TITLE
Model optimization by integrating growth rate data from biolog

### DISCUSSION THREAD

**Discussion by hites77** [opencobra/cobrapy#1390] [Q&A]
Is there any way to optimize the model by integrating growth rate data from Biolog plate ?

**Comment by cdiener** [comment]
Absolutely. The growth rate is the objective and can be constrained as you like. One gotcha is that you would need to scale the import/exchange fluxes correctly to match your growth conditions because they are in units of $\frac{mmol}{gDW\cdot h}$ but your growth medium will probably be in the units of $mmol/L$ or $g/L$. So you would have to make sure it's correctly transform to 1gDW of cells and getting an estimate for the length of the steady state to divide by. If this is not possible you could also just predict growth rates and use the measured ones as validation.

Constraining the growth rate:

```python
from cobra.io import load_model

mod = load_model(textbook)
print(mod.objective.expression)  # find out what is optimized

# Constrain growth rate to 0.5/h
mod.reactions.Biomass_Ecoli_core.upper_bound = 0.5
sol = mod.optimize()
print(mod.summary(solution=sol))
```

That will respect the bound:

```
Objective
=========
1.0 Biomass_Ecoli_core = 0.5

Uptake
------
Metabolite    Reaction  Flux  C-Number  C-Flux
  glc__D_e EX_glc__D_e    10         6 100.00%
     nh4_e    EX_nh4_e 2.726         0   0.00%
      o2_e     EX_o2_e 38.14         0   0.00%
      pi_e     EX_pi_e 1.839         0   0.00%

Secretion
---------
Metabolite Reaction   Flux  C-Number  C-Flux
     co2_e EX_co2_e -38.72         1 100.00%
     h2o_e EX_h2o_e -42.36         0   0.00%
       h_e   EX_h_e -10.03         0   0.00%

```

**Reply by hites77** [reply]
Thanks for initiating the response to my query. The back calculation seems to be a possible option. So, for different carbon source the calculation needs to be carried out while adding that carbon in the model. I think it will be generating multiple flux solution for multiple reactions (if multiple carbon sources are considered from a Biolog PM plate) that will need a way for optimization. Am i thinking in correct direction ?

**Reply by cdiener** [reply]
Yes exactly, you would have different media for different carbon sources and different growth rate constraints for each sample.
